lib/src/code_field/code_controller.dart

```diff
@@ -334,6 +324,141 @@ class CodeController extends TextEditingController {
     );
   }
 
+  void outdentSelection() {
+    final tabSpaces = params.tabSpaces;
+    if (selection.start == -1 || selection.end == -1) {
+      return;
+    }
+
+    modifySelectedLines((line) {
+      if (line == '\n') {
+        return line;
+      }
+
+      if (line.length < tabSpaces) {
+        return line.trimLeft();
+      }
+
+      final subStr = line.substring(0, tabSpaces);
+      if (subStr == ' ' * tabSpaces) {
+        return line.substring(tabSpaces, line.length);
+      }
+      return line.trimLeft();
+    });
+  }
+
+  void indentSelection() {
+    final tabSpaces = params.tabSpaces;
+    final tab = ' ' * tabSpaces;
+    final lines = _code.lines.lines;
+    if (selection.start == -1 || selection.end == -1) {
+      return;
+    }
+
+    if (selection.isCollapsed) {
+      final fullPosition = _code.hiddenRanges.recoverPosition(
+        selection.start,
+        placeHiddenRanges: TextAffinity.downstream,
+      );
+      final lineIndex = _code.lines.characterIndexToLineIndex(fullPosition);
+      final columnIndex = fullPosition - lines[lineIndex].textRange.start;
+      final insert = ' ' * (tabSpaces - (columnIndex % tabSpaces));
+      insertStr(insert);
+      return;
+    }
+
+    modifySelectedLines((line) {
+      if (line == '\n') {
+        return line;
+      }
+      return tab + line;
+    });
+  }
+
+  /// Filters the lines that have at least one character selected.
+  ///
+  /// IMPORTANT: this method also changes the selection to be:
+  /// start: start of the first selected line
+  /// end: end of the last line
+  ///
+  /// Folded blocks are considered to be selected
+  /// if they are located between start and end of a selection.
+  ///
+  /// [modifierCallback] - transformation function that modifies the line.
+  /// `line` in the callback contains '\n' symbol at the end, except for the last one.
+  void modifySelectedLines(String Function(String line) modifierCallback) {
```

Comment:
Maybe we should return the changed `TextEditingValue` instead of changing it here? I have a slight feeling that this is cleaner but cannot explain well.